### DISCUSSION TITLE
Are Server side API routes shared or unique to each user instance?

### DISCUSSION THREAD

**Discussion by entropygeneration** [vercel/next.js#51599] [Help] [answered]
### Summary

Was unable to find any information clarifying how api routes are utilized when launching to vercel. I was curious to know if when creating server side api routes, if those routes become shared by all users of the application, or if each user of the application has their own instance of the api route? Any clarity or insight appreciated.

**Comment by icyJoseph** [comment, marked as answer]
On Vercel API routes (Netlify and others which support serverless out of the box), and most any server side work is done as tasks, which are pretty much serverless functions.

Once a request arrives, if no task active for that work, one is spun up (in the edge there's near zero wake up time), and the task does its job. Then after some time, if no more requests arrive, the task is terminated. 

- https://youtu.be/jv4Soz08gBw?t=1270

In that video, each block has a border coloured, using a unique identifier generated by the task at `/api/zone` - all the images are rendered by the same endpoint, but many tasks have been triggered! He calls them lambdas.

So yeah, you should not make assumptions, it might be that 1, 2, 3,... users share a task, so you need to keep them stateless. Also, they'll be terminated, so you can't really store much JS context information, nor are you allowed to manipulate the file system. Other than through the tmp folder: https://vercel.com/guides/how-can-i-use-files-in-serverless-functions#using-temporary-storage - which is only useful to temporarily store data that you'll send elsewhere.

**Reply by entropygeneration** [reply]
Thank you very much for the detailed response, really helps clarify.
